### PR TITLE
React-powered SVG exports

### PR DESCRIPTION
**Pull request by SomeHats**
## Migration path
1. If any of your shapes implement `toSvg` for exports, you'll need to replace your implementation with a new version that returns JSX (it's a react component) instead of manually constructing SVG DOM nodes
2. `editor.getSvg` is deprecated. It still works, but will be going away in a future release. If you still need SVGs as DOM elements rather than strings, use `new DOMParser().parseFromString(svgString, 'image/svg+xml').firstElementChild`

## The change in detail
At the moment, our SVG exports very carefully try to recreate the visuals of our shapes by manually constructing SVG DOM nodes. On its own this is really painful, but it also results in a lot of duplicated logic between the `component` and `getSvg` methods of shape utils.

In #3020, we looked at using string concatenation & DOMParser to make this a bit less painful. This works, but requires specifying namespaces everywhere, is still pretty painful (no syntax highlighting or formatting), and still results in all that duplicated logic.

I briefly experimented with creating my own version of the javascript language that let you embed XML like syntax directly. I was going to call it EXTREME JAVASCRIPT or XJS for short, but then I noticed that we already wrote the whole of tldraw in this thing called react and a (imo much worse named) version of the javascript xml thing already existed. 

Given the entire library already depends on react, what would it look like if we just used react directly for these exports? Turns out things get a lot simpler! Take a look at lmk what you think

This diff was intended as a proof of concept, but is actually pretty close to being landable. The main thing is that here, I've deliberately leant into this being a big breaking change to see just how much code we could delete (turns out: lots). We could if we wanted to make this without making it a breaking change at all, but it would add back a lot of complexity on our side and run a fair bit slower

### CODE REVIEW

**Comment by steveruizok** [comment]
Surprised that all of our SVG stuff was only 2.7k LOC!

What are the breaking changes between this and what we had before?

**Comment by steveruizok** [comment]
this is sick

**Comment by si14** [comment]
My understanding is SVG export was a bottleneck for @TodePond, and I suppose this might make exports slower?

HOWEVER, apparently TS [can emit whatever](https://www.typescriptlang.org/tsconfig#jsxFactory) instead of `React.createElement`, so what if we wrap everything in our EXTREME NOTREACT (EXTRACT for short) not-elements and serialise that, sidestepping DOM and VDOM entirely?

**Comment by steveruizok** [comment]
I'd like to test the speed, though I can't really tell the difference myself.

I have a small concern that this is a further step toward rendering only in React, though I can live with that. (Maybe the next library will be renderer agnostic).

**Comment by TodePond** [comment]
> My understanding is SVG export was a bottleneck for @TodePond, and I suppose this might make exports slower?

I believe that @SomeHats didn't go for the approach I was warning about in the end, so maybe ok?

**Comment by SomeHats** [comment]
> What are the breaking changes between this and what we had before?
1. I removed a bunch of helpers we had around converting SVG DOM nodes to strings as we can just generate the strings directly. If you're using our higher level helpers you're fine
2. The `getSvg` that returns a DOM node is still there but i think i might deprecate it?
3. Anyone who's written their own `toSvg` on a shape will get errors if they upgrade to this - it'll still be returning DOM nodes, where we're now expecting JSX. Upgrade path is converting that to JSX, but afaik these export things aren't documented so idk if many people are actually using them

> I'd like to test the speed, though I can't really tell the difference myself.
In my own experimentation: this is a bit slower, but not by much. Doing SVG -> image (not touched by this) is typically 3-4x slower than generating the SVG, and for any exports that include fonts or assets then fetching those massively dominates the export time (even locally it's another 3-4x on the base export time). I can do some proper benchmarking and come up with some real numbers if folks would like though!

> HOWEVER, apparently TS [can emit whatever](https://www.typescriptlang.org/tsconfig#jsxFactory) instead of React.createElement, so what if we wrap everything in our EXTREME NOTREACT (EXTRACT for short) not-elements and serialise that, sidestepping DOM and VDOM entirely?
This would be fun, but probably isn't a good idea for a couple of reasons:
1. we are actually making heavy use of react features in our export code now - primarily context, because we're re-using the react components we use for displaying the actual shapes
2. library users would need to adopt our custom compiler or special typescript modifications
3. there's no way i know of to differentiate between two flavours of JSX in the same project - we'd need to do something funky to get that to only apply where we wanted it
4. dom creation/serialisation isn't the bottle-neck on exports anyway (see above)

in practice, `renderToStaticMarkup` is probably sufficient here: we will create some extra objects to model the JSX, but no actual DOM nodes will get created we you take that approach.

**Comment by si14** [comment]
@SomeHats I see, thank you!

1 and 4 are probably deal breakers, but fwiw 2 and 3 are easy: we can split SVG-y bits into separate `.tsx` files and ask TS to emit the right thing by placing `/** @jsx hacketyHack */` on top of those `.tsx` files. No modifications, no custom compilers, not very funky, just vibes

**Comment by mimecuvalo** [comment]
just chiming in to say that this looks wonderful 😊

**Comment by SomeHats** [comment]
> Surprised that all of our SVG stuff was only 2.7k LOC!
Found another 600 lines 😅

**Review by ds300 — COMMENTED** [review]
This is huge, well done! Leaving my comments here for the weekend but I still have some unresolved Qs for myself to check on monday so will do a bit more review then.

File: apps/examples/e2e/tests/export-snapshots.spec.ts-snapshots/Export-snapshots-Exports-with-note-font-draw-in-dark-mode-1-Mobile-Chrome-linux.png
[the remark this answers is not in the document]
Comment (ds300) [reply apps/examples/e2e/tests/export-snapshots.spec.ts-snapshots/Export-snapshots-Exports-with-note-font-draw-in-dark-mode-1-Mobile-Chrome-linux.png:1]: I tested this out manually and couldn't see any contrast drop so 🤷🏼 probably fine to ignore.

File: apps/dotcom/setupTests.js [hunk 1]
```diff
@@ -2,3 +2,6 @@ global.crypto ??= new (require('@peculiar/webcrypto').Crypto)()
 
 process.env.MULTIPLAYER_SERVER = 'https://localhost:8787'
 process.env.ASSET_UPLOAD = 'https://localhost:8788'
+
+global.TextEncoder = require('util').TextEncoder
+global.TextDecoder = require('util').TextDecoder
```
Comment (ds300) [inline apps/dotcom/setupTests.js:7]: what are these for? does renderToStaticMarkup use them?

File: packages/editor/src/lib/components/default-components/DefaultCanvas.tsx [hunk 2]
```diff
@@ -499,32 +499,39 @@ function DebugSvgCopy({ id }: { id: TLShapeId }) {
...
 			latest = null
 			unsubscribe()
 		}
 	}, [editor, id, isInRoot])
+	const bb = editor.getShapePageBounds(id)
 
-	if (!isInRoot) return null
+	if (!isInRoot || !src || !bb) return null
 
 	return (
```
Comment (ds300) [inline packages/editor/src/lib/components/default-components/DefaultCanvas.tsx:522]: the debug mode seems to not work at all in this PR. it's currently broken in staging too, but only by the svg offsets being out of whack. Here the svgs don't seems to show up at all?

File: packages/editor/src/lib/editor/getSvgJsx.tsx [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+		// Expand by an extra 32 pixels
+		bbox.expandBy(padding)
+	}
+
+	// We want the svg image to be BIGGER THAN USUAL to account for image quality
+	const w = bbox.width * scale
+	const h = bbox.height * scale
+
+	try {
+		document.body.focus?.() // weird but necessary
```
[the remark this answers is not in the document]
Comment (steveruizok) [reply packages/editor/src/lib/editor/getSvgJsx.tsx:68]: Might not be necessary?

File: packages/editor/src/lib/components/default-components/DefaultCanvas.tsx (hunk 2 above)
Comment (SomeHats) [reply packages/editor/src/lib/components/default-components/DefaultCanvas.tsx:522]: huh, this was some weird css we had in .com so i wasn't seeing this issue on the examples app. fixed now!

File: packages/tldraw/src/lib/shapes/frame/FrameShapeUtil.tsx [hunk 4]
```diff
@@ -97,93 +93,97 @@ export class FrameShapeUtil extends BaseBoxShapeUtil<TLFrameShape> {
...
-		textBg.setAttribute('height', `${opts.height}px`)
-		textBg.setAttribute('rx', 4 + 'px')
-		textBg.setAttribute('ry', 4 + 'px')
-		textBg.setAttribute('fill', theme.background)
-
-		g.appendChild(textBg)
-		g.appendChild(text)
-
-		return g
+		// TODO: text label
```
[the remark this answers is not in the document]
Comment (SomeHats) [reply packages/tldraw/src/lib/shapes/frame/FrameShapeUtil.tsx:112]: fixed! thanks for the reminder :)

File: packages/editor/src/lib/primitives/utils.ts [hunk 5]
```diff
@@ -339,17 +339,17 @@ export function pointInPolygon(A: VecLike, points: VecLike[]): boolean {
  * @public
  */
 export function toDomPrecision(v: number) {
-	return +v.toFixed(4)
```
[the remark this answers is not in the document]
Comment (TodePond) [reply packages/editor/src/lib/primitives/utils.ts:342]: wow nice

**Pull request merged**